lib/class-wp-webfonts.php

```diff
@@ -0,0 +1,323 @@
+			$this->stylesheet_handle = 'webfonts-footer';
+			$hook                    = 'wp_print_footer_scripts';
+		} else {
+			$this->stylesheet_handle = 'webfonts';
+			$hook                    = 'wp_enqueue_scripts';
+		}
+		add_action( $hook, array( $this, 'generate_and_enqueue_styles' ) );
+
+		// Enqueue webfonts in the block editor.
+		add_action( 'admin_init', array( $this, 'generate_and_enqueue_editor_styles' ) );
+	}
+
+	/**
+	 * Get the list of fonts.
+	 *
+	 * @return array
+	 */
+	public function get_fonts() {
+		return self::$webfonts;
+	}
+
+	/**
+	 * Get the list of providers.
+	 *
+	 * @return array
+	 */
+	public function get_providers() {
+		return self::$providers;
+	}
+
+	/**
+	 * Register a webfont.
+	 *
+	 * @param array $font The font arguments.
+	 */
+	public function register_font( $font ) {
+		$font = $this->validate_font( $font );
+		if ( $font ) {
+			$id                    = $this->get_font_id( $font );
+			self::$webfonts[ $id ] = $font;
+		}
+	}
+
+	/**
+	 * Get the font ID.
+	 *
+	 * @param array $font The font arguments.
+	 * @return string
+	 */
+	public function get_font_id( $font ) {
+		return sanitize_title( "{$font['font-family']}-{$font['font-weight']}-{$font['font-style']}-{$font['provider']}" );
+	}
+
+	/**
+	 * Validate a font.
+	 *
+	 * @param array $font The font arguments.
+	 *
+	 * @return array|false The validated font arguments, or false if the font is invalid.
+	 */
+	public function validate_font( $font ) {
+		$font = wp_parse_args(
+			$font,
+			array(
+				'provider'     => 'local',
+				'font-family'  => '',
+				'font-style'   => 'normal',
+				'font-weight'  => '400',
+				'font-display' => 'fallback',
+			)
+		);
+
+		// Check the font-family.
+		if ( empty( $font['font-family'] ) || ! is_string( $font['font-family'] ) ) {
+			trigger_error( __( 'Webfont font family must be a non-empty string.', 'gutenberg' ) );
+			return false;
+		}
+
+		// Local fonts need a "src".
+		if ( 'local' === $font['provider'] ) {
+			// Make sure that local fonts have 'src' defined.
+			if ( empty( $font['src'] ) || ( ! is_string( $font['src'] ) && ! is_array( $font['src'] ) ) ) {
+				trigger_error( __( 'Webfont src must be a non-empty string or an array of strings.', 'gutenberg' ) );
+				return false;
+			}
+		}
+
+		// Validate the 'src' property.
+		if ( ! empty( $font['src'] ) ) {
+			foreach ( (array) $font['src'] as $src ) {
+				if ( empty( $src ) || ! is_string( $src ) ) {
+					trigger_error( __( 'Each webfont src must be a non-empty string.', 'gutenberg' ) );
+					return false;
+				}
+
+				if (
+					// Validate data URLs.
+					! preg_match( '/^data:.+;base64/', $src ) &&
+					// Validate URLs.
+					! filter_var( $src, FILTER_VALIDATE_URL ) &&
+					// Check if it's a URL starting with "//" (omitted protocol).
+					0 !== strpos( $src, '//' )
+				) {
+					trigger_error( __( 'Webfont src must be a valid URL or a data URI.', 'gutenberg' ) );
+					return false;
+				}
```

Comment:
Frankly don't think this is needed. It checks if the developer has hard-coded an URL in the webfonts settings. Do we really think a developer will not add an URL there? And even if they don't what's the point in running this slower check again and again? :) 

Lets try to figure out how this would work. 
1. With this code in place:
    i. If the check passes and the string is indeed an URL first it will pass the `preg_match()` test, then it will pass the `filter_var()` test and finally will pass the `strpos()` test. These three tests will run on every page load (that has webfont) "forever". The change any of them would fail on a particular page load is practically near zero as the webfonts settings are typically hard-coded.
    ii. Lets assume that one if the three tests fail. What happens is that an PHP error will be triggered that is supposed to write a line in the PHP error log. Then registration of this webfont is halted and all of the HTML for it will be missing from the webpage. If that happens in production there will be no warnings or any kind of indication why the webfont is not working. The site admin will have to access the PHP error logs to be able to even see that there's been an error. 
2. Without this code:
    i. No tests are run on the URL. No delays.
    ii. Assuming the string is not a valid URL, the webfont HTML will still be outputted to the browser and there will be an error in the browser tools/console. In production the site admins will be able to easily see that something is wrong and report the error to the theme/plugin author.

Seems that number 2 is a lot more desirable behaviour and will make it easier to find and report errors for both developers and users.

Reply:
Fair point. Removed 👍